### PR TITLE
Added multithreading for download process

### PR DESCRIPTION
**Pull request by zuenko**
Now it is much faster.

### CODE REVIEW

**Comment by zuenko** [comment]
@keldrom, If you want to use increase number of threads use --n_jobs default - 4
Also, I saw that labels creation not very effective. Still figuring out how to optimize it.
```
[INFO] Creating labels for Car of train.
 12%|███▊                            | 10628/89465 [3:56:58<29:17:50,  1.34s/it]
```

**Comment by keldrom** [comment]
I will try the code in the next few days and then I will update all. Thanks btw for your great work.

**Comment by zuenko** [comment]
Hello again! 

I fixed get_label function. Why simple .csv? Because we all have different tools, for example, I like YOLO and it doesn't accept the string label name, and also I need to get height and width of a box. So for that, I will need to load every .txt file change and save it. 

So, why don't we just save all labels into .csv and people will proceed it as they want. And for addition, increase execution time.

**Comment by EscVM** [comment]
Hi @zuenko !

I've tested the multithreading feature, that you added with the first commit, and it seems working pretty well. It's much more faster and efficient. I have only a couple of questions: how many threads can be added and why four is the default one? I've tried with 50 and is considerably faster so, what's the limit?

Regarding the second commit, I totally agree that a .csv file can satisfactory fix the terribly slow process of creating all the .txt files. It still not totally clear to me why that happens.

Anyway, I'm still on vacation and so, I still don't have time to work on the three points that you have pointed out:
- Multi-class creation with a single .csv
- Fix the Visualization tool
- Fast making .txt label, prototype. For this last point, it could be cool to add the possibility to specify the wanted .txt layout. In this way, we could already give the possibility to obtain the desired format for what so ever model will be used.
If you want and have time to fix these points I'll merge your work with the master branch.

Thank you again for your help and interest to this project.

**Comment by zuenko** [comment]
Hello @EscVM !

Yes, I confirm that 4 threads not so effective as it can be because I have a very good performing PC. Also, I've changed '--n_jobs' to '--threads', and will add soon that we can use up to 200 threads.

> Regarding the second commit, I totally agree that a .csv file can satisfactorily fix the terribly slow process of creating all the .txt files. It still not totally clear to me why that happens.

I think that exceptions has a significant influence on performance. So I applied slicing pandas with checking labels in the directory. 

Here is example for making .txt for yolo:
https://gist.github.com/zuenko/80bfbba7b0d34b30a9e5589b4c8d01a8
353220/353220 [5:30:40<00:00, 17.80it/s] - averege about 18 it/s without depending on Pool or threads.

To sum up, I will try to fix all points in the near future, glad to help!

**Comment by zuenko** [comment]
What was done: 

- [x] Fix visualizer
- [x] Fix multiclass
- [x] Fix defaults
- [x] Add example
- [x] One feature for visualizer, if you at the end of the list, return to start.

**Comment by borisd13** [comment]
@zuenko , I tested your pull request and I get a lot of error messages after running
`python main.py download --classes Flower --type_csv all`

No file is actually downloaded.

See attached
aws_not_found

**Comment by zuenko** [comment]
@borisd13 

Hello @borisd13, thank you!

But [awscli](https://github.com/zuenko/OIDv4_ToolKit/blob/master/requirements.txt#L3) is inside requirements.txt. For the second case, yes, seems logical I will rename the folder.

**Comment by keldrom** [comment]
@zuenko Multiclass is not only the merging of the folders but the creation of a .txt that contains all the needed classes; for our studies was useful and we think that it can be the same for other usages.
Btw we will check if the code runs as you said and then we will add your great work to the project.

**Comment by keldrom** [comment]
@borisd13 `modules` is a folder that contains all the codes used in `main.py`

**Comment by zuenko** [comment]
@keldrom I don't agree with it can be useful for you, but for other more simply give an examples for different _Format_ and than they should decide. 
As I mentioned before, labels need to be more universal, so I changed the tool for doing to the simple _.csv_. 
For the second point, you do creation _.txt_ in the single class as well and if multiclass == 1 does:
[https://github.com/zuenko/OIDv4_ToolKit/blob/master/main.py#L124](Loop)
for each class and putting all images in the same folder. If we switch to .csv annotation we don't need that argument. But we can rewrite algorithm to put in the same directory classes which can cross in the single image and that will be perfect.

**Comment by EscVM** [comment]
@zuenko using a .csv file, as I already told you, is a great idea
(especially for speed improvement).

Instead, multiclass was thought for cases like the following one: for
example, if you want to download chairs and tables, there will be for sure
images where both classes appear at the same time. So, for those pictures,
it would be better to have both annotations and train the model with all
instances annotated at the same time. I hope I am clear in this.
Thank you again for your help.

**Comment by EscVM** [comment]
Anyway, now I came back from my holiday so, I have time again to follow the
project ;)

**Comment by EscVM** [comment]
Hi @zuenko! I tried your commit and I found some "problems":

- with threads is really faster than before. However, I still don't get the point to select the number of threads. I've tried with 200 and it still perfectly works. So, why don't remove from the options the multi-threading feature and leave it behind the curtains? Leave for example 20 or 30 that can be changed only through the code
- in the visualization I still don't totally like the menu. I'll think about it and I'll change it by myself
- finally, and this one is the most important point, instead of using a jupyter notebook as example to create .txt files, I would integrate the creation of them in the code. I would leave the possibility, after the download of the images and creation of the labels, to **exit** or to **create the .txt** files. It would be perfect to integrate that functionality with a menu that let you select the desired format ( I want all .txt with Xmin Xmax Ymin Ymax or Xmax Ymax Xmin Ymin ...).

Ones all these problems are fixed, we'll merge all your branch with the master one.

**Comment by borisd13** [comment]
@zuenko @keldrom I retested using pipenv and can confirm it works correctly. Please disregard my previous comments.

**Comment by EscVM** [comment]
So, @zuenko, what do you think?

**Comment by zuenko** [comment]
@EscVM , Sorry, very intensive week, I have some thoughts about it. 

> with threads is really faster than before. However, I still don't get the point to select the number of threads. I've tried with 200 and it still perfectly works. So, why don't remove from the options the multi-threading feature and leave it behind the curtains? Leave for example 20 or 30 that can be changed only through the code

For this point, I think we should give people choose because we already have the default number if they want it can be changed without going into the code.

> in the visualization I still don't totally like the menu. I'll think about it and I'll change it by myself

When I was testing my code, I just really tired of typing again and again validation. The menu with nums makes it more simple. 

> finally, and this one is the most important point, instead of using a jupyter notebook as example to create .txt files, I would integrate the creation of them in the code. I would leave the possibility, after the download of the images and creation of the labels, to exit or to create the .txt files. It would be perfect to integrate that functionality with a menu that let you select the desired format ( I want all .txt with Xmin Xmax Ymin Ymax or Xmax Ymax Xmin Ymin ...).

Before that, we should think about some questions:

- How to encode labels? (Before my commit you were putting the name in str format, it may cause problems for users)
- Wich formats should we implement?
- How to optimize the speed?
- How to sort them?

So the main challenge, for now, it the balance between complexity of choice and simplifying it. 

And for now, we can try to add to options for encoding, yolo, classic and .csv. 

Currently, I'm very busy but will try to help you as fast as I can. Good luck!

**Comment by EscVM** [comment]
@zuenko no problem. You have already done a lot of things.

> For this point, I think we should give people choose because we already have the default number if they want it can be changed without going into the code. 

I can't understand what it means increasing it to 200. Why can I increase that parameter to so higher values?

> And for now, we can try to add to options for encoding, yolo, classic and .csv.

Yes, it's the best option.

**Comment by EscVM** [comment]
Changes:

- [x] fix the visualizer menu
- [x] add multi-threading feature made by @zuenko 
- [x] update Readme with the new changes
